File: examples/src/main/python/autoencoder_bench.py
"""Steady-state cost of one autoencoder training step in plain JAX, JIT-compiled and eager.

The reference this benchmarks against is
`examples/src/main/scala/deepwit/examples/autoencoder/AutoEncoderBench.scala`, which measures
the DeepWit program of the same shape under the same protocol. Equivalence is structural,
not just numerical: the model is written as a per-sample function under `jax.vmap`, the
reconstruction loss as a per-pixel function under a second `jax.vmap`, and the optimizer as
DeepWit's `Adam` transcribed term for term — because that is what DeepWit's `vmap`, `zipvmap`
and `Adam` lower to. A batched matmul formulation would measure a different program.

Run, from the repository root:
    uv run --no-sync python examples/src/main/python/autoencoder_bench.py --steps 100 --reps 10
"""

from __future__ import annotations

import argparse
import json
import pickle
import statistics
import struct
import time
from pathlib import Path
from typing import Callable, NamedTuple

import jax
import jax.numpy as jnp
import numpy as np

REPO_ROOT = Path(__file__).resolve().parents[4]
EXAMPLES_ROOT = REPO_ROOT / "examples"

# -- Configuration, mirroring AutoEncoderTrain / AutoEncoderBench --

BATCH_SIZE = 256
LATENT_DIM = 24
LEARNING_RATE = 3e-4
E_HIDDEN1 = 512
E_HIDDEN2 = 256
D_HIDDEN1 = 256
D_HIDDEN2 = 512
INPUT_SIZE = 28 * 28

BETA1 = 0.9
BETA2 = 0.999
EPSILON = 1e-8


# -- Parameters, one node per case class in Autoencoder.Params --


class AffineParams(NamedTuple):
    weight: jax.Array
    bias: jax.Array


class EncoderParams(NamedTuple):
    layer1: AffineParams
    layer2: AffineParams
    latent_layer: AffineParams


class DecoderParams(NamedTuple):
    layer1: AffineParams
    layer2: AffineParams
    output_layer: AffineParams


class Params(NamedTuple):
    encoder: EncoderParams
    decoder: DecoderParams


class AdamState(NamedTuple):
    momentums: Params
    velocities: Params
    beta1t: jax.Array
    beta2t: jax.Array


class TrainState(NamedTuple):
    params: Params
    optimizer_state: AdamState
    last_cost: jax.Array


# -- Model: `y = xW + b`, exactly what deepwit's AffineLayer contracts to --


def affine(p: AffineParams, x: jax.Array) -> jax.Array:
    return x.dot(p.weight) + p.bias


def encode(p: EncoderParams, x: jax.Array) -> jax.Array:
    h1 = jax.nn.relu(affine(p.layer1, x))
    h2 = jax.nn.relu(affine(p.layer2, h1))
    return affine(p.latent_layer, h2)


def decode(p: DecoderParams, z: jax.Array) -> jax.Array:
    h1 = jax.nn.relu(affine(p.layer1, z))
    h2 = jax.nn.relu(affine(p.layer2, h1))
    return affine(p.output_layer, h2)


def logits(params: Params, x: jax.Array) -> jax.Array:
    return decode(params.decoder, encode(params.encoder, x))


def bce_from_logits(target: jax.Array, logit: jax.Array) -> jax.Array:
    """`max(z, 0) - z*y + log(1 + exp(-|z|))`, as deepwit's BinaryCrossEntropy writes it."""
    return jnp.maximum(logit, 0.0) - logit * target + jnp.log(1.0 + jnp.exp(-jnp.abs(logit)))


def cost_fn_for(samples: jax.Array) -> Callable[[Params], jax.Array]:
    def cost(params: Params) -> jax.Array:
        def per_sample(sample: jax.Array) -> jax.Array:
            original = sample.reshape(-1)
            return jax.vmap(bce_from_logits)(original, logits(params, original)).sum()

        return jax.vmap(per_sample)(samples).mean()

    return cost


# -- Adam, transcribed from dimwit.optimizer.Adam --


def adam_init(params: Params) -> AdamState:
    # Two distinct zero trees, as dimwit's `def zeros` gives: sharing one tree between
    # momentums and velocities would hand the donating JIT step the same buffer twice.
    def zeros() -> Params:
        return jax.tree.map(jnp.zeros_like, params)

    return AdamState(zeros(), zeros(), jnp.float32(1.0), jnp.float32(1.0))


def adam_update(grads: Params, params: Params, state: AdamState) -> tuple[Params, AdamState]:
    m = jax.tree.map(lambda m_prev, g: BETA1 * m_prev + (1.0 - BETA1) * g, state.momentums, grads)
    v = jax.tree.map(lambda v_prev, g: BETA2 * v_prev + (1.0 - BETA2) * g**2, state.velocities, grads)
    beta1t = state.beta1t * BETA1
    beta2t = state.beta2t * BETA2
    m_hat = jax.tree.map(lambda x: x / (1.0 - beta1t), m)
    v_hat = jax.tree.map(lambda x: x / (1.0 - beta2t), v)
    new_params = jax.tree.map(
        lambda p, mh, vh: p - (LEARNING_RATE * mh) / (jnp.sqrt(vh) + EPSILON), params, m_hat, v_hat
    )
    return new_params, AdamState(m, v, beta1t, beta2t)


def gradient_step(batch: jax.Array, state: TrainState) -> TrainState:
    cost, grads = jax.value_and_grad(cost_fn_for(batch))(state.params)
    new_params, new_optimizer_state = adam_update(grads, state.params, state.optimizer_state)
    return TrainState(new_params, new_optimizer_state, cost)


jit_gradient_step = jax.jit(gradient_step, donate_argnums=(1,))


# -- Data, read the same way MNISTLoader reads it --


def load_train_images(data_dir: Path) -> jax.Array:
    with open(data_dir / "train-images-idx3-ubyte", "rb") as f:
        magic, count, rows, cols = struct.unpack(">IIII", f.read(16))
        assert magic == 2051, f"Invalid magic: {magic}"
        assert (rows, cols) == (28, 28), f"Invalid image dimensions: {rows} x {cols}"
        pixels = np.frombuffer(f.read(), dtype=np.uint8).reshape(count, rows, cols)
    return jnp.asarray(pixels, dtype=jnp.float32) / 255.0


def init_params(key: jax.Array) -> Params:
    """Glorot uniform with zero bias, which is what AffineLayer.Params.init does.

    Only used when no DeepWit-exported parameters are supplied; the benchmark loads
    those by preference so that both programs start from the same point.
    """

    def layer(k: jax.Array, fan_in: int, fan_out: int) -> AffineParams:
        limit = np.sqrt(6.0 / (fan_in + fan_out))
        return AffineParams(
            weight=jax.random.uniform(k, (fan_in, fan_out), minval=-limit, maxval=limit),
            bias=jnp.zeros((fan_out,), dtype=jnp.float32),
        )

    keys = jax.random.split(key, 6)
    return Params(
        encoder=EncoderParams(
            layer(keys[0], INPUT_SIZE, E_HIDDEN1),
            layer(keys[1], E_HIDDEN1, E_HIDDEN2),
            layer(keys[2], E_HIDDEN2, LATENT_DIM),
        ),
        decoder=DecoderParams(
            layer(keys[3], LATENT_DIM, D_HIDDEN1),
            layer(keys[4], D_HIDDEN1, D_HIDDEN2),
            layer(keys[5], D_HIDDEN2, INPUT_SIZE),
        ),
    )


def load_exported_params(path: Path) -> Params:
    """Read the parameters DeepWit's TensorTreeIO pickled: nested tuples of numpy arrays."""
    with open(path, "rb") as f:
        (encoder, decoder) = pickle.load(f)

    def group(layers) -> list[AffineParams]:
        return [AffineParams(jnp.asarray(w), jnp.asarray(b)) for (w, b) in layers]

    return Params(encoder=EncoderParams(*group(encoder)), decoder=DecoderParams(*group(decoder)))


# -- Measurement, the same protocol as AutoEncoderBench.measure --


def measure(
    label: str,
    step: Callable[[jax.Array, TrainState], TrainState],
    batch: jax.Array,
    initial: Callable[[], TrainState],
    warmup: int,
    reps: int,
    steps_per_rep: int,
) -> dict:
    # A thunk, not a state: the donating JIT step deletes the buffers it is handed,
    # so every measurement has to be given a state of its own.
    state = jax.block_until_ready(initial())

    first_call_start = time.perf_counter()
    state = jax.block_until_ready(step(batch, state))
    first_call_ms = (time.perf_counter() - first_call_start) * 1e3

    for _ in range(warmup):
        state = step(batch, state)
    state = jax.block_until_ready(state)

    step_ms_per_rep = []
    for _ in range(reps):
        start = time.perf_counter()
        for _ in range(steps_per_rep):
            state = step(batch, state)
        state = jax.block_until_ready(state)
        step_ms_per_rep.append((time.perf_counter() - start) * 1e3 / steps_per_rep)

    assert not np.isnan(state.last_cost), f"{label} diverged to NaN"

    mean = statistics.fmean(step_ms_per_rep)
    return {
        "label": label,
        "first_call_ms": first_call_ms,
        "steps_per_rep": steps_per_rep,
        "reps": reps,
        "mean_step_ms": mean,
        "std_step_ms": statistics.stdev(step_ms_per_rep) if reps > 1 else 0.0,
        "median_step_ms": statistics.median(step_ms_per_rep),
        "min_step_ms": min(step_ms_per_rep),
        "max_step_ms": max(step_ms_per_rep),
        "steps_per_second": 1000.0 / mean,
        "step_ms_per_rep": step_ms_per_rep,
    }


def loss_trace(batch: jax.Array, initial: TrainState, steps: int) -> list[float]:
    state = initial
    trace = []
    for _ in range(steps):
        state = jit_gradient_step(batch, state)
        trace.append(float(state.last_cost))
    return trace


def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--steps", type=int, default=100, help="JIT steps per timed repetition")
    parser.add_argument("--eager-steps", type=int, default=20, help="eager steps per timed repetition")
    parser.add_argument("--reps", type=int, default=10, help="timed repetitions")
    parser.add_argument("--warmup", type=int, default=20, help="discarded steps before timing")
    parser.add_argument("--trace", type=int, default=20, help="steps of loss trace to record")
    parser.add_argument("--mode", choices=("both", "jit", "eager"), default="both")
    parser.add_argument("--out", type=Path, default=EXAMPLES_ROOT / "out" / "bench")
    parser.add_argument("--data", type=Path, default=EXAMPLES_ROOT / "data")
    args = parser.parse_args()

    args.out.mkdir(parents=True, exist_ok=True)

    device = jax.devices()[0]
    print(f"Device: {device}")
    print(f"JAX:    {jax.__version__}")

    images = load_train_images(args.data)
    batch = jax.block_until_ready(images[:BATCH_SIZE])

    exported = args.out / "init_params.pkl"
    if exported.exists():
        params = load_exported_params(exported)
        print(f"Parameters: loaded from {exported}")
    else:
        params = init_params(jax.random.key(42))
        print("Parameters: freshly initialized (no DeepWit export found)")

    def initial() -> TrainState:
        # A fresh copy each time: the donating JIT step deletes the buffers it is handed.
        fresh = jax.tree.map(lambda x: x.copy(), params)
        return TrainState(fresh, adam_init(fresh), jnp.float32(-1.0))

    batch_checksum = float(batch.sum())
    print(f"Batch checksum (sum of pixels): {batch_checksum:.6f}")

    # Measured before the loss trace, so that the first call of each configuration is a
    # genuinely cold compile rather than one served from a cache the trace has already filled.
    measurements = []
    if args.mode in ("both", "jit"):
        measurements.append(
            measure("jax-jit", jit_gradient_step, batch, initial, args.warmup, args.reps, args.steps)
        )
    if args.mode in ("both", "eager"):
        measurements.append(
            measure("jax-eager", gradient_step, batch, initial, args.warmup, args.reps, args.eager_steps)
        )

    print()
    header = f"{'configuration':<28} {'mean ms':>9}   {'std':<8} {'median':>9} {'min':>9} {'max':>9} {'steps/s':>9} {'1st call ms':>10}"
    print(header)
    for m in measurements:
        print(
            f"{m['label']:<28} {m['mean_step_ms']:9.3f} ± {m['std_step_ms']:<8.3f} "
            f"{m['median_step_ms']:9.3f} {m['min_step_ms']:9.3f} {m['max_step_ms']:9.3f} "
            f"{m['steps_per_second']:9.1f} {m['first_call_ms']:10.1f}"
        )

    trace = loss_trace(batch, jax.block_until_ready(initial()), args.trace)
    with open(args.out / "python_loss_trace.csv", "w") as f:
        f.write("step,loss\n")
        for i, loss in enumerate(trace, start=1):
            f.write(f"{i},{loss}\n")
    print(f"Loss trace: {trace[0]} .. {trace[-1]}")

    report = {
        "implementation": "jax-python",
        "device": str(device),
        "jax_version": jax.__version__,
        "batch_size": BATCH_SIZE,
        "latent_dim": LATENT_DIM,
        "learning_rate": LEARNING_RATE,
        "warmup_steps": args.warmup,
        "batch_checksum": batch_checksum,
        "measurements": measurements,
    }
    with open(args.out / "python_bench.json", "w") as f:
        json.dump(report, f, indent=2)
    print(f"\nWrote {args.out / 'python_bench.json'}")


if __name__ == "__main__":
    main()

File: examples/src/main/scala/deepwit/examples/autoencoder/AutoEncoderBench.scala
package deepwit.examples.autoencoder

import dimwit.*
import dimwit.Conversions.given
import dimwit.jax.Jax
import dimwit.optimizer.Adam
import dimwit.tensortree.{TensorTree, TensorTreeIO}

import deepwit.examples.dataset.MNISTLoader
import deepwit.loss.BinaryCrossEntropy

import java.nio.charset.StandardCharsets
import java.nio.file.{Files, Path}

/** Steady-state cost of one autoencoder training step, JIT-compiled and eager.
  *
  * The timed region is exactly one `gradientStep` — value-and-grad of the batch cost,
  * then the Adam update — on a batch that is already resident on the device. Loading
  * and batching sit outside it. `examples/src/main/python/autoencoder_bench.py` runs the
  * same protocol over a hand-written JAX program of the same shape, so the two numbers
  * are comparable.
  *
  * Run as, from the repository root:
  * {{{
  * sbt "examples/runMain deepwit.examples.autoencoder.benchAutoEncoder steps=100 reps=10"
  * }}}
  */
object AutoEncoderBench:

  trait Batch derives Label

  // -- Configuration, mirroring AutoEncoderTrain --

  val batchSize = 256
  val latentDim = 24
  val learningRate = 3e-4f

  val eHidden1Extent = Axis[EHidden1] -> 512
  val eHidden2Extent = Axis[EHidden2] -> 256
  val latentExtent = Axis[Latent] -> latentDim
  val dHidden1Extent = Axis[DHidden1] -> 256
  val dHidden2Extent = Axis[DHidden2] -> 512

  val optimizer = Adam(learningRate = learningRate)

  // -- The step under measurement, character for character the one in AutoEncoderTrain --

  def costFnFor[S: Label](samples: Tensor3[S, Height, Width, Float32])(params: Autoencoder.Params): Tensor0[Float32] =
    val model = Autoencoder(params)
    samples
      .vmap(Axis[S]): sample =>
        val original = sample.flatten
        zipvmap(Axis[Pixel])(original, model.logits(original)): (origPixel, reconPixel) =>
          BinaryCrossEntropy.fromLogits(origPixel, reconPixel)
        .sum
      .mean

  def gradientStep(batch: Tensor3[Batch, Height, Width, Float32], state: TrainState): TrainState =
    val (cost, grads) = Autodiff.valueAndGrad(costFnFor(batch))(state.params)
    val (newParams, newOptimizerState) = optimizer.update(grads, state.params, state.optimizerState)
    TrainState(newParams, newOptimizerState, cost)

  def initialState(key: Key): TrainState =
    val params = Autoencoder.Params.init(eHidden1Extent, eHidden2Extent, latentExtent, dHidden1Extent, dHidden2Extent, key)
    TrainState(params, optimizer.init(params), Tensor0(-1f))

  // -- Measurement --

  /** JAX dispatch is asynchronous, so a timed region only means something once every
    * array it produced has actually landed.
    */
  def blockUntilReady[T: TensorTree](tree: T): T =
    Jax.jax.block_until_ready(TensorTree[T].toPyTree(tree))
    tree

  case class Measurement(label: String, firstCallMs: Double, stepMsPerRep: Seq[Double], stepsPerRep: Int):
    private val sorted = stepMsPerRep.sorted
    val reps: Int = stepMsPerRep.size
    val mean: Double = stepMsPerRep.sum / reps
    val std: Double =
      if reps < 2 then 0.0
      else math.sqrt(stepMsPerRep.map(t => (t - mean) * (t - mean)).sum / (reps - 1))
    val median: Double = if reps % 2 == 1 then sorted(reps / 2) else (sorted(reps / 2 - 1) + sorted(reps / 2)) / 2
    val min: Double = sorted.head
    val max: Double = sorted.last
    val stepsPerSecond: Double = 1000.0 / mean

    def render: String =
      f"$label%-28s ${mean}%9.3f ± ${std}%-8.3f ${median}%9.3f ${min}%9.3f ${max}%9.3f ${stepsPerSecond}%9.1f ${firstCallMs}%10.1f"

    def toJson: String =
      s"""    {
         |      "label": "$label",
         |      "first_call_ms": $firstCallMs,
         |      "steps_per_rep": $stepsPerRep,
         |      "reps": $reps,
         |      "mean_step_ms": $mean,
         |      "std_step_ms": $std,
         |      "median_step_ms": $median,
         |      "min_step_ms": $min,
         |      "max_step_ms": $max,
         |      "steps_per_second": $stepsPerSecond,
         |      "step_ms_per_rep": [${stepMsPerRep.mkString(", ")}]
         |    }""".stripMargin

  /** One measurement: a discarded first call, `warmup` discarded steps, then `reps`
    * timed runs of `stepsPerRep` steps each. Each run is blocked on at its end, inside
    * the timed region, so no work escapes the clock.
    *
    * `initial` is by-name because the donating JIT step deletes the buffers it is handed,
    * so every measurement has to be given a state of its own.
    */
  def measure[B, S: TensorTree](
      label: String,
      step: (B, S) => S,
      batch: B,
      initial: => S,
      lastCost: S => Tensor0[Float32],
      warmup: Int,
      reps: Int,
      stepsPerRep: Int
  ): Measurement =
    var state = blockUntilReady(initial)

    val firstCallStart = System.nanoTime()
    state = blockUntilReady(step(batch, state))
    val firstCallMs = (System.nanoTime() - firstCallStart) / 1e6

    for _ <- 1 to warmup do state = step(batch, state)
    state = blockUntilReady(state)

    val stepMsPerRep =
      for _ <- 1 to reps yield
        val start = System.nanoTime()
        for _ <- 1 to stepsPerRep do state = step(batch, state)
        state = blockUntilReady(state)
        (System.nanoTime() - start) / 1e6 / stepsPerRep

    // Keep the trajectory honest: a diverged run would have been measuring NaNs.
    val finalCost = lastCost(state).item
    require(!finalCost.isNaN, s"$label diverged to NaN")

    Measurement(label, firstCallMs, stepMsPerRep, stepsPerRep)

  /** The loss over the first `steps` steps, for checking the Python program computes
    * the same thing rather than merely something of the same shape.
    */
  def lossTrace[B, S](step: (B, S) => S, batch: B, initial: => S, lastCost: S => Tensor0[Float32], steps: Int): Seq[Float] =
    var state = initial
    for _ <- 1 to steps yield
      state = step(batch, state)
      lastCost(state).item

  /** Cost of marshalling one state out to Python and back again.
    *
    * This is what a JIT call pays on every invocation on top of running the executable:
    * `toPyTree` of the arguments on the way in, `fromPyTree` of the result on the way out.
    */
  def roundTripMs[S: TensorTree](state: S, iterations: Int): Double =
    val tree = TensorTree[S]
    for _ <- 1 to 20 do tree.fromPyTree(tree.toPyTree(state))
    val start = System.nanoTime()
    for _ <- 1 to iterations do tree.fromPyTree(tree.toPyTree(state))
    (System.nanoTime() - start) / 1e6 / iterations

  /** How many tensors a state is made of, which is what the marshalling cost scales with. */
  def leafCount[S: TensorTree](state: S): Int =
    TensorTree[S].mapLeaves(state, [T <: Tuple, V] => (labels: Labels[T]) ?=> (x: Tensor[T, V]) => 1).sum

@main
def benchAutoEncoder(args: String*): Unit =
  import AutoEncoderBench.*

  val opts = args.map(_.split("=", 2)).collect { case Array(k, v) => k -> v }.toMap
  val stepsPerRep = opts.getOrElse("steps", "100").toInt
  val reps = opts.getOrElse("reps", "10").toInt
  val warmup = opts.getOrElse("warmup", "20").toInt
  val eagerStepsPerRep = opts.getOrElse("eagerSteps", "20").toInt
  val traceSteps = opts.getOrElse("trace", "20").toInt
  val outDir = Path.of(opts.getOrElse("out", "out/bench"))
  val modes = opts.getOrElse("mode", "both")

  Files.createDirectories(outDir)

  val device = Jax.devices.head
  println(s"Device: $device")
  println(s"JAX:    ${Jax.jax.__version__}")

  // -- The batch under measurement: the first batch of the training stream, on device --

  val trainDataset = MNISTLoader.createTrainingDataset().get
  val batch = trainDataset.toBatchStream(Axis[Batch] -> batchSize).map(_.images).next()
  blockUntilReady(batch)

  val initKey = Key(42)

  // Same starting point for the Python program, so the loss traces are comparable.
  TensorTreeIO.save(initialState(initKey).params, outDir.resolve("init_params.pkl"))

  val batchChecksum = batch.sum.item
  println(f"Batch checksum (sum of pixels): $batchChecksum%.6f")

  // Measured before the loss trace, so that the first call of each configuration is a
  // genuinely cold compile rather than one served from a cache the trace has already filled.
  val measurements =
    val jitted =
      if modes == "eager" then Seq.empty
      else
        val jitStep = jitDonatingUnsafe(gradientStep)
        Seq(measure("deepwit-jit", jitStep, batch, initialState(initKey), _.lastCost, warmup, reps, stepsPerRep))
    val eager =
      if modes == "jit" then Seq.empty
      else Seq(measure("deepwit-eager", gradientStep, batch, initialState(initKey), _.lastCost, warmup, reps, eagerStepsPerRep))
    jitted ++ eager

  println()
  println(f"${"configuration"}%-28s ${"mean ms"}%9s   ${"std"}%-8s ${"median"}%9s ${"min"}%9s ${"max"}%9s ${"steps/s"}%9s ${"1st call ms"}%10s")
  measurements.foreach(m => println(m.render))

  val json =
    s"""{
       |  "implementation": "deepwit",
       |  "device": "$device",
       |  "jax_version": "${Jax.jax.__version__}",
       |  "batch_size": $batchSize,
       |  "latent_dim": $latentDim,
       |  "learning_rate": $learningRate,
       |  "warmup_steps": $warmup,
       |  "batch_checksum": ${batchChecksum.toDouble},
       |  "measurements": [
       |${measurements.map(_.toJson).mkString(",\n")}
       |  ]
       |}
       |""".stripMargin
  val trace = lossTrace(jitDonatingUnsafe(gradientStep), batch, initialState(initKey), _.lastCost, traceSteps)
  Files.writeString(
    outDir.resolve("scala_loss_trace.csv"),
    ("step,loss" +: trace.zipWithIndex.map { case (l, i) => s"${i + 1},${l.toDouble}" }).mkString("\n") + "\n",
    StandardCharsets.UTF_8
  )
  println(s"Loss trace: ${trace.head} .. ${trace.last}")

  Files.writeString(outDir.resolve("deepwit_bench.json"), json, StandardCharsets.UTF_8)
  println(s"\nWrote ${outDir.toAbsolutePath}/deepwit_bench.json")

/** The six-layer autoencoder swept over batch size.
  *
  * The depth sweep moves work and leaf count together. Batch size moves only the work: the
  * parameter tree is the same 39 tensors at every point here, so DeepWit's per-call bridge cost
  * is a constant that a larger batch has more computation to hide behind. This is the axis on
  * which the two implementations should converge.
  *
  * Run, from the repository root:
  * {{{
  * sbt "examples/runMain deepwit.examples.autoencoder.benchAutoEncoderBatch batches=64,256,1024,4096,16384"
  * }}}
  */
@main
def benchAutoEncoderBatch(args: String*): Unit =
  import AutoEncoderBench.*

  val opts = args.map(_.split("=", 2)).collect { case Array(k, v) => k -> v }.toMap
  val batchSizes = opts.getOrElse("batches", "64,256,1024,4096,16384").split(",").map(_.trim.toInt).toSeq
  val stepsPerRep = opts.getOrElse("steps", "100").toInt
  val reps = opts.getOrElse("reps", "10").toInt
  val warmup = opts.getOrElse("warmup", "20").toInt
  val eagerStepsPerRep = opts.getOrElse("eagerSteps", "5").toInt
  val eagerReps = opts.getOrElse("eagerReps", "8").toInt
  val outDir = Path.of(opts.getOrElse("out", "out/bench"))
  val modes = opts.getOrElse("mode", "both")

  Files.createDirectories(outDir)

  val device = Jax.devices.head
  println(s"Device: $device")
  println(s"JAX:    ${Jax.jax.__version__}")

  val trainDataset = MNISTLoader.createTrainingDataset().get
  val initKey = Key(42)

  // Measured once: the parameter tree does not depend on the batch, which is the point.
  val probe = blockUntilReady(initialState(initKey))
  val leaves = leafCount(probe)
  val bridgeMs = roundTripMs(probe, 100)
  println(f"Train state: $leaves%d tensors, bridge round trip ${bridgeMs}%.3f ms")

  // Same starting point for the Python program, so the two are comparable numerically.
  TensorTreeIO.save(initialState(initKey).params, outDir.resolve("init_params.pkl"))

  val results =
    for batchSize <- batchSizes yield
      val batch = blockUntilReady(trainDataset.toBatchStream(Axis[Batch] -> batchSize).map(_.images).next())
      println(s"\n== batch $batchSize ==")
      val measurements =
        val jitted =
          if modes == "eager" then Seq.empty
          else
            val jitStep = jitDonatingUnsafe(gradientStep)
            Seq(measure(s"deepwit-jit-b$batchSize", jitStep, batch, initialState(initKey), _.lastCost, warmup, reps, stepsPerRep))
        val eager =
          if modes == "jit" then Seq.empty
          else Seq(measure(s"deepwit-eager-b$batchSize", gradientStep, batch, initialState(initKey), _.lastCost, warmup, eagerReps, eagerStepsPerRep))
        jitted ++ eager
      measurements.foreach(m => println(m.render))
      Jax.gc()
      (batchSize, measurements)

  println()
  println(f"${"configuration"}%-28s ${"mean ms"}%9s   ${"std"}%-8s ${"median"}%9s ${"min"}%9s ${"max"}%9s ${"steps/s"}%9s ${"1st call ms"}%10s")
  results.foreach(_._2.foreach(m => println(m.render)))

  val configsJson = results.map: (batchSize, measurements) =>
    s"""    {
       |      "batch_size": $batchSize,
       |      "measurements": [
       |${measurements.map(_.toJson).mkString(",\n")}
       |      ]
       |    }""".stripMargin

  val json =
    s"""{
       |  "implementation": "deepwit",
       |  "model": "autoencoder-6layer",
       |  "device": "$device",
       |  "jax_version": "${Jax.jax.__version__}",
       |  "latent_dim": ${AutoEncoderBench.latentDim},
       |  "learning_rate": ${AutoEncoderBench.learningRate},
       |  "warmup_steps": $warmup,
       |  "leaves": $leaves,
       |  "round_trip_ms": $bridgeMs,
       |  "configs": [
       |${configsJson.mkString(",\n")}
       |  ]
       |}
       |""".stripMargin
  Files.writeString(outDir.resolve("deepwit_batch_bench.json"), json, StandardCharsets.UTF_8)
  println(s"\nWrote ${outDir.toAbsolutePath}/deepwit_batch_bench.json")
